include/tvm/runtime/vm.h

```diff
@@ -0,0 +1,420 @@
+ */
+struct VMFunction {
+  /*! \brief The function's name. */
+  std::string name;
+  /*! \brief The number of function parameters. */
+  Index params;
+  /*! \brief The instructions representing the function. */
+  std::vector<Instruction> instructions;
+  /*! \brief The size of the frame for this function */
+  Index register_file_size;
+
+  VMFunction(std::string name, Index params, std::vector<Instruction> instructions,
+             Index register_file_size)
+      : name(name),
+        params(params),
+        instructions(instructions),
+        register_file_size(register_file_size) {}
+
+  VMFunction() {}
+
+  friend std::ostream& operator<<(std::ostream& os, const VMFunction&);
+};
+
+/*! \brief A representation of a stack frame.
+ *
+ * A stack frame is a record containing the information needed
+ * to restore the caller's virtual machine state after returning
+ * from a function call.
+ */
+struct VMFrame {
+  /*! \brief The return program counter. */
+  Index pc;
+  /*! \brief The index into the function table, points to the caller. */
+  Index func_index;
+  /*! \brief The number of arguments. */
+  Index args;
+  /*! \brief A pointer into the caller function's instructions. */
+  const Instruction* code;
+
+  /*! \brief Statically allocated space for objects */
+  std::vector<Object> register_file;
+
+  /*! \brief Register in caller's frame to put return value */
+  RegName caller_return_register;
+
+  VMFrame(Index pc, Index func_index, Index args, const Instruction* code, Index register_file_size)
+      : pc(pc),
+        func_index(func_index),
+        args(args),
+        code(code),
+        register_file(register_file_size),
+        caller_return_register(0) {}
+};
+
+/*! \brief The virtual machine.
+ *
+ * The virtual machine contains all the current execution state,
+ * as well as the global view of functions, the global constant
+ * table, the compiled operators.
+ *
+ * The goal is to have a single self-contained object,
+ * enabling one to easily pass around VMs, execute them on
+ * multiple threads, or serialized them to disk or over the
+ * wire.
+ */
+struct VirtualMachine {
+  /*! \brief The virtual machine's packed function table. */
+  std::vector<PackedFunc> packed_funcs;
+  /*! \brief The virtual machine's function table. */
+  std::vector<VMFunction> functions;
+  /*! \brief The current stack of call frames. */
+  std::vector<VMFrame> frames;
+  /*! \brief The global constant pool. */
+  std::vector<Object> constants;
+  /*! \brief The fuction table index of the current function. */
+  Index func_index;
+  /*! \brief The current pointer to the code section. */
+  const Instruction* code;
+  /*! \brief The virtual machine PC. */
+  Index pc;
+
+  /*! \brief The special return register. */
+  Object return_register;
+
+  /*! \brief The set of TVM contexts the VM is currently executing on. */
+  std::vector<TVMContext> ctxs;
+
+  /*! \brief Push a call frame on to the frame stack. */
```

Comment:
```suggestion
  /*! \brief Push a call frame on to the call stack. */
```

To be consistent with terminology.